### PR TITLE
[GR-31691] Initial JFR support.

### PR DESCRIPTION
**Pull request by christianhaeubl**

### CODE REVIEW

**Comment by christianhaeubl** [comment]
@jiekang : instead of a rebase, I used a more selective approach. This changed the hashes of the following RedHat commits: 8f220de, 9bbfeaf, and 4e53401. All other commits are either my own or they should have the same hash as in your `jfr-backup` branch.

**Comment by jiekang** [comment]
> * Hello Jie Kang, thanks for contributing a PR to our project!
> 
> We use the [Oracle Contributor Agreement](http://www.oracle.com/technetwork/community/oca-486395.html) to make the copyright of contributions clear. We don't have a record of you having signed this yet, based on your email address jkang -(at)- redhat -(dot)- com. You can sign it at that link.
> 
> If you think you've already signed it, please comment below and we'll check.
> 
> * Hello Victor Ma, thanks for contributing a PR to our project!
> 
> We use the [Oracle Contributor Agreement](http://www.oracle.com/technetwork/community/oca-486395.html) to make the copyright of contributions clear. We don't have a record of you having signed this yet, based on your email address vma -(at)- redhat -(dot)- com. You can sign it at that link.
> 
> If you think you've already signed it, please comment below and we'll check.

Both Jie and Victor (jkang@redhat.com and vma@redhat.com) are Red Hat employees and I believe have the OCA signed on their behalf.

**Comment by jiekang** [comment]
> @jiekang : instead of a rebase, I used a more selective approach. This changed the hashes of the following RedHat commits: [8f220de](https://github.com/oracle/graal/commit/8f220ded26491644a022fd6b790cd999e1684a46), [9bbfeaf](https://github.com/oracle/graal/commit/9bbfeafad350f2897d99bb8aede699bd581fdb9b), and [4e53401](https://github.com/oracle/graal/commit/4e534012b7158a4dee3ab37f5eca6698b4a5226f). All other commits are either my own or they should have the same hash as in your `jfr-backup` branch.

Okay. I think the commit contents and authors look fine. Thank you.

File: substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceId.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * current JFR epoch.
+ */
+public class JfrTraceId {
+
+    public static final long BIT = 1;
+    public static final long META_SHIFT = 8;
+    public static final long TRANSIENT_META_BIT = (BIT << 3);
+    public static final long TRANSIENT_BIT = (TRANSIENT_META_BIT << META_SHIFT);
+    public static final long SERIALIZED_META_BIT = (BIT << 4);
+    public static final long SERIALIZED_BIT = (SERIALIZED_META_BIT << META_SHIFT);
```
Comment (teshull) [inline substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceId.java:47]: It doesn't seem to me like `TRANSIENT_BIT` or `SERIALIZED_BIT` can ever be set within a tag.

If so, then these could be deleted, as well as the checks for the `TRANSIENT_BIT`

Comment (rkennke) [reply substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceId.java:47]: Right, those are unused in this incarnation. We might need them later, when (and if) we implement stuff like leak detection, but should probably add them back only when we actually need them. They can be safely removed.

File: substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceIdLoadBarrier.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ * iterate over them e.g. when writing the constant pools to disk.
+ */
+public class JfrTraceIdLoadBarrier {
+    private static final UninterruptibleUtils.AtomicInteger classCount0 = new UninterruptibleUtils.AtomicInteger(0);
+    private static final UninterruptibleUtils.AtomicInteger classCount1 = new UninterruptibleUtils.AtomicInteger(0);
+
+    @Uninterruptible(reason = "Epoch may not change")
+    private static boolean isNotTagged(long value) {
+        long thisEpochBit = JfrTraceIdEpoch.getInstance().thisEpochBit();
+        return ((value & ((thisEpochBit << JfrTraceId.META_SHIFT) | thisEpochBit)) != thisEpochBit);
```
[the remark this answers is not in the document]
Comment (rkennke) [reply substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceIdLoadBarrier.java:45]: It checks if the Class is tagged  as used in the current epoch (as opposed to the previous epoch).

[the remark this answers is not in the document]
Comment (teshull) [reply substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceIdLoadBarrier.java:45]: where is the value shifted? I see, `JftTraceIdEpoch.changeEpoch()`, but is cleanup done somewhere else as well?

[the remark this answers is not in the document]
Comment (christianhaeubl) [reply substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceIdLoadBarrier.java:45]: I will remove this method and do a few more cleanups as those things are no longer needed anyways.

**Pull request merged**